### PR TITLE
Add a map renderer

### PR DESCRIPTION
**Pull request by DrSmugleaf**
## About the PR

Adds:
- Content.MapRenderer, a project that takes a map file and renders it out onto an image. It supports doing multiple maps.
- ~~A GitHub action that shows changes done to maps..~~

GitHub action TBD in another PR.

(Outdated) Current run time locally: 35 second client+server startup time, 8 second painting time.

**Screenshots**
saltern

**TODOs**

- [ ] Making it run ~in 40 milliseconds~ faster (it takes me 35 seconds to run it locally, 2 minutes in CI and it only does one map file at a time for example).
- [x] Icon smoothing.
- [x] Layer direction.
- [ ] RSI delays and animations.
- [ ] Adding support for selectively rendering entities (including/excluding wires, including container contents).
- [ ] Separating these different versions of the map into different images.
- [x] Hiding the dummy client entity.
- [ ] Removing the submodule update.
- [ ] Make it take entities into accounts for image bounds, not just the grid box.
- [ ] Test with multiple grids.
- [x] Pinging PJB to edit the repo's environment variables to add a github token so we can have the action directly commit instead of opening a PR.
- [x] PJB doing the thing.
- [ ] ~~Make it use os mesa and llvmpipe for off screen rendering. (when PJB compiles it)~~
- [ ] Properly rendering entities that aren't 32x32

### CODE REVIEW

File: .github/workflows/map-diff.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
+name: Map Diff
+
+on:
+  push:
+#    branches:
+#    - master
```
Comment (DrSmugleaf) [inline .github/workflows/map-diff.yml:6]: Someone remind me to change this before it gets merged ty

**Comment by SweptWasTaken** [comment]
This is sexy and I love it and I can't wait for it. Excellent work!!
nedladdning_10

**Comment by SweptWasTaken** [comment]
Also if it isn't suggested or implemented yet, the bot should post a direct viewable link to the file somewhere when it runs the action thingy.

Like:

Here are the changes for this map! - <https://github.com/DrSmugleaf/space-station-14/blob/20a16577dae2aedb42c7e70f934b562f826e2c73/Resources/MapImages/saltern.png>

Also what if I modify two maps in the commit? Is there support for that?

**Comment by DrSmugleaf** [comment]
> Also what if I modify two maps in the commit? Is there support for that?
Yes, it will run for every created and modified map in a commit to master, however many.

> Also if it isn't suggested or implemented yet, the bot should post a direct viewable link to the file somewhere when it runs the action thingy.
It doesn't currently, I'll look into doing that though.

**Comment by moonheart08** [comment]
Is it possible to run this on the server normally, to take full map screenshots?

**Comment by DrSmugleaf** [comment]
> Is it possible to run this on the server normally, to take full map screenshots?

Not unless you run the map differ separately, independent of whether or not you have the game started up. It could probably be modified to do that though the current method would take around 8 seconds to paint Saltern.

**Comment by moonheart08** [comment]
Really, being able to take map screenshots at all would be nice, even if it is slow. The other solution would be a screenshot-based move-the-camera magic trick, which'd be good for pretty screenshots, but has it's own issues.

**Comment by DrSmugleaf** [comment]
> Really, being able to take map screenshots at all would be nice, even if it is slow. The other solution would be a screenshot-based move-the-camera magic trick, which'd be good for pretty screenshots, but has it's own issues.

I tried to do that originally but it wouldn't work for CI without os mesa/llvmpipe fuckery, which is out of scope for this PR.
I'll see about making a command for it in a future PR though.

**Comment by SweptWasTaken** [comment]
If this PR gets merged, i'll map, PR and maintain /vg/ Boxstation immediately after.

**Comment by DrSmugleaf** [comment]
THE RETURN OF THE KING

**Review by metalgearsloth — APPROVED** [review]
probably fine xd

File: a.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+A
```
Comment (metalgearsloth) [inline a.txt:1]: A

Comment (moonheart08) [reply a.txt:1]: A

**Comment by SweptWasTaken** [comment]
https://github.com/DrSmugleaf/space-station-14/commit/d5985d2a2dc361de5c8cd1d3ade558a6bd0c848d

File: a.txt (hunk 2 above)
Comment (ZeroDayDaemon) [reply a.txt:1]: A

Comment (DrSmugleaf) [reply a.txt]: [A](https://github.com/space-wizards/space-station-14/pull/3613/commits/428a29c695f97f0d8e5369fd0580faa7b4eef222)

**Comment by DrSmugleaf** [comment]
finished-elijah-wood

**Pull request merged**